File: 04-ISP/cpp/bad_design.cpp
#include <iostream>
#include <stdexcept>

class Machine {
public:
    virtual void printDocument() = 0;
    virtual void scanDocument() = 0;
    virtual void faxDocument() = 0;

    virtual ~Machine() = default;
};

class BasicPrinter : public Machine {
public:
    void printDocument() override {
        std::cout << "Printing document\n";
    }

    void scanDocument() override {
        throw std::runtime_error("BasicPrinter cannot scan");
    }

    void faxDocument() override {
        throw std::runtime_error("BasicPrinter cannot fax");
    }
};

int main() {
    BasicPrinter printer;

    printer.printDocument();

    return 0;
}

File: 04-ISP/python/bad_design.py
from abc import ABC, abstractmethod

class Machine(ABC):

    @abstractmethod
    def print_document(self):
        pass

    @abstractmethod
    def scan_document(self):
        pass

    @abstractmethod
    def fax_document(self):
        pass


class BasicPrinter(Machine):

    def print_document(self):
        print("Printing document")

    def scan_document(self):
        raise NotImplementedError("BasicPrinter cannot scan")

    def fax_document(self):
        raise NotImplementedError("BasicPrinter cannot fax")

if __name__ == "__main__":
    printer = BasicPrinter()

    printer.print_document()

'''
PS C:\Users\Shikha Pandey\Documents\github\solid-design-principles>python 04-ISP/python/bad_design.py"
Printing document
'''